cpp/src/rolling/window_utils.cu

```diff
@@ -0,0 +1,255 @@
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else if (window_type == window_type::RIGHT_CLOSED) {
+      thrust::copy_n(rmm::exec_policy(stream),
+                     cudf::detail::make_counting_transform_iterator(
+                       0,
+                       distance_kernel<T, OffsetType, op_t<Which, window_type::RIGHT_CLOSED>>{
+                         d_length, d_offset, input_begin, input_end}),
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else if (window_type == window_type::CLOSED) {
+      thrust::copy_n(rmm::exec_policy(stream),
+                     cudf::detail::make_counting_transform_iterator(
+                       0,
+                       distance_kernel<T, OffsetType, op_t<Which, window_type::CLOSED>>{
+                         d_length, d_offset, input_begin, input_end}),
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else {
+      CUDF_FAIL("Unhandled window type.");
+    }
+    return result;
+  }
+
+  template <typename T, CUDF_ENABLE_IF(cudf::is_timestamp<T>())>
+  [[nodiscard]] std::unique_ptr<column> operator()(column_view const& input,
+                                                   scalar const& length,
+                                                   scalar const& offset,
+                                                   window_type const window_type,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using OffsetType = typename T::duration;
+    using ScalarType = duration_scalar<OffsetType>;
+    CUDF_EXPECTS(cudf::is_duration(length.type()), "Length and offset must be duration types.");
+    CUDF_EXPECTS(length.type().id() == type_to_id<OffsetType>(),
+                 "Length must have same the resolution as the input.");
+    return compute_window_bounds<T, OffsetType, ScalarType>(
+      input, length, offset, window_type, stream, mr);
+  }
+
+  template <typename T, CUDF_ENABLE_IF(cudf::is_index_type<T>() and !cudf::is_unsigned<T>())>
+  [[nodiscard]] std::unique_ptr<column> operator()(column_view const& input,
+                                                   scalar const& length,
+                                                   scalar const& offset,
+                                                   window_type const window_type,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using OffsetType = T;
+    using ScalarType = numeric_scalar<OffsetType>;
+    CUDF_EXPECTS(have_same_types(input, length),
+                 "Input column, length, and offset must have the same type.");
```

Comment:
Similarly here, I think it would be OK if `length` and `offset` are addable to `input`. But this is easier.